### PR TITLE
Only make input paths absolute when -working-directory is passed

### PR DESCRIPTION
**Pull request by rmaz**
This is a partial revert of https://github.com/swiftlang/swift-driver/commit/a9f35bc15f68d4ae0cd3949afa6c78e1a3df8f80

When remotely building and sharing Swift artifacts it is important to maintain
relative paths to avoid serializing any paths that will not be valid on different
machines. Change the logic back to the previous model, and the one used
in the old swift driver, where paths are made absolute only when a 
`-working-directory` flag is passed.

### CODE REVIEW

**Comment by rmaz** [comment]
@swift-ci please test

**Review by cachemeifyoucan — COMMENTED** [review]
The direction of the change works for me. Forgot distributed build is using the relative path to do the compilation.

Note there is an alternative to use `-scanner-prefix-map` to alter the path returned from the scanner but I guess that option requires knowing the path on the remote node before hand.

File: Sources/SwiftDriver/Driver/Driver.swift [hunk 1]
```diff
@@ -688,14 +689,14 @@ public struct Driver {
...
+      let cwd = fileSystem.currentWorkingDirectory
       return try cwd.map{ try AbsolutePath(validating: workingDirectoryArg.asSingle, relativeTo: $0) } ?? AbsolutePath(validating: workingDirectoryArg.asSingle)
-    } ?? cwd
-
-    // Apply the working directory to the parsed options.
-    if let workingDirectory = self.workingDirectory {
-      try Self.applyWorkingDirectory(workingDirectory, to: &self.parsedOptions)
+    }
+    
+    if let specifiedWorkingDir = self.workingDirectory {
```
[the remark this answers is not in the document]
Comment (rmaz) [reply Sources/SwiftDriver/Driver/Driver.swift:697]: I tried that initially, but then the output paths (and profile data paths) end up absolute. I didn't track down where that was happening, but there seems to be multiple places in the code that would end up prefixing paths if `workingDirectory` was set.

File: Tests/SwiftDriverTests/SwiftDriverTests.swift [hunk 2]
```diff
@@ -45,12 +45,15 @@ private var testInputsPath: AbsolutePath {
   }
 }
 
-func toPath(_ path: String, base: AbsolutePath = localFileSystem.currentWorkingDirectory!) throws -> VirtualPath {
-  return try VirtualPath(path: path).resolvedRelativePath(base: base)
+func toPath(_ path: String, isRelative: Bool = true) throws -> VirtualPath {
```
[the remark this answers is not in the document]
Comment (rmaz) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:48]: > I am slightly prefer leaving the tests in the new state but add -working-directory to the driver command.

For the changed tests? For ` testDOTFileEmission` that would work, for the others it would require adding `isRelative: false` to all the `toPath()` calls, which would be a bigger change. 

> Also, would be nice to add a test to check the behavior of path arguments with/without -working-directory

I added `testRelativeInputs` on L7239 for this

File: Sources/SwiftDriver/Driver/Driver.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (cachemeifyoucan) [reply Sources/SwiftDriver/Driver/Driver.swift:697]: Maybe driver will think that working-directory is passed via command-line so it is passing working directory (in abs path) frontend flag plus some other flags derived from working directory?

Sounds fine with me.

**Review by cachemeifyoucan — APPROVED** [review]
Thanks for fixing!

**Comment by rmaz** [comment]
@swift-ci please test

**Comment by rmaz** [comment]
@swift-ci Please test Windows platform

File: Tests/SwiftDriverTests/SwiftDriverTests.swift [hunk 3]
```diff
@@ -3039,10 +3040,10 @@ final class SwiftDriverTests: XCTestCase {
     XCTAssertEqual(emitModuleJob.outputs[0].file, try toPath("Test.swiftmodule"))
     XCTAssertEqual(emitModuleJob.outputs[1].file, try toPath("Test.swiftdoc"))
     XCTAssertEqual(emitModuleJob.outputs[2].file, try toPath("Test.swiftsourceinfo"))
-    XCTAssertEqual(emitModuleJob.outputs[3].file, try toPath("Test.swiftinterface"))
+    XCTAssertEqual(emitModuleJob.outputs[3].file, try VirtualPath(path: "./Test.swiftinterface"))
     XCTAssertEqual(emitModuleJob.outputs[4].file, try toPath("Test.private.swiftinterface"))
     XCTAssertEqual(emitModuleJob.outputs[5].file, try toPath("Test-Swift.h"))
-    XCTAssertEqual(emitModuleJob.outputs[6].file, try toPath("Test.tbd"))
+    XCTAssertEqual(emitModuleJob.outputs[6].file, try VirtualPath(path: "./Test.tbd"))
```
Comment (artemcm) [inline Tests/SwiftDriverTests/SwiftDriverTests.swift:3046]: @rmaz do you recall why this change was needed for only these specific outputs?

Comment (rmaz) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:3046]: I don't, but changing it to `toPath()` gives us:
```
XCTAssertEqual failed: ("./Test.swiftinterface") is not equal to ("Test.swiftinterface")
```
So somewhere a CWD of `.` is being prefixed only on the `swiftinterface` path and the `tbd` path.

Comment (rmaz) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:3046]: After some code wandering, it is because the path is unspecified in the compiler args, so falls back to this codepath to determine it:
https://github.com/swiftlang/swift-driver/blob/eba19529226c95a602bfd1cecaef3ca920bb199f/Sources/SwiftDriver/Driver/Driver.swift#L3623-L3626

This ends up calling `dirname` on a relative path with no folder components, which ends up here:
https://github.com/swiftlang/swift-tools-support-core/blob/b464fcd8d884e599e3202d9bd1eee29a9e504069/Sources/TSCBasic/Path.swift#L598-L605

This ends up calculating the parent as `.` and generates the filename as `modulename.swiftinterface`.

Comment (artemcm) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:3046]: Thank you!
